### ISSUE TITLE
Regarding Spelling Error model

### ISSUE THREAD

**Issue by smilenrhyme** [deepmipt/DeepPavlov#964]
Thanks for amazing toolkit :) Can you please share your views on below questions

1. How does **correct_prior** & **incorrect_prior** calculation done in Error model ?
2. How do we incorporate "**count**" with incorrect-correct pair e.g. if training data is in form of (intended_word, observed_word, count).
3. Is there any other way we can combine LM score & EM score in LM beam search method ?

Thanks a lot !!

**Comment by yoptar** [comment]
Hi @smilenrhyme,
Thank you for your kind comment
1. The priors were chosen empirically when trying to maximize accuracy on the [kartaslov](https://github.com/dkulagin/kartaslov/blob/master/dataset/orfo_and_typos/orfo_and_typos.L1_5.csv) dataset for Russian language
1. The simplest way, I think, would be to just repeat the pair that many times.
1. One could try and add some constant scaling to EM probabilities. Or train a simple ranking model that would do the same but automatically. We tweaked the probabilities for the simple Levenshtein model but not the BrillMoore one.

**Comment by smilenrhyme** [comment]
Thanks for quick response :) Some more points 

1. Do you mean maximizing accuracy on End-to-End Spell task w.r.t above point 1 ? Or measure Error model in isolation ?
2. Just to clarify, even with "occurance_count" equation would be same because  p = c / e (ignoring priors) so p = (c * occurance_count) / (e * occurance_count). So the "p" value remains same. Is this correct ?

In addition, 
3. Getting edit distance = 3 for 'invetar' <-> 'inverter' while training (where we are splitting word into chars with spaces) otherwise if we call directly "_distance_edits" its 2. Is that correct ? and any reason for not to consider words with more than 2-edit in training.

Thanks once again !!

**Comment by smilenrhyme** [comment]
Hi @yoptar, Could you please share your thoughts for above points. It would be very helpful for me. Thanks a lot.

**Comment by yoptar** [comment]
Hi @smilenrhyme, sorry for the late response.
1. I meant error model in isolation
2. Yes, occurrence count by itself does not change the p value, but since we are talking about characters and character groups, `c` and `e` ration will most likely change, because the same characters and character groups also occur in other words
3. I didn't understand about splitting words in chars, I'm not sure it should happen. We skipped errors with more that 2-edit because in the kartaslov dataset most of those words are really weird, they would just mess up the probabilities.

**Comment by smilenrhyme** [comment]
@yoptar Thanks a lot once again :)

1. Could you please suggest some metric to measure error model accuracy in isolation (Task independent).
2. It clears out my doubts 👍 
3. For example : 
    `print(model._distance_edits("invetar", "inverter"))`  --> **2**
    `print(model._distance_edits("⟬i n v e t a r⟭", "⟬i n v e r t e r⟭"))`   --> **3**  [It got filters out in error_model .fit() method as we are ingoing pairs with more than 2 edit. ]
  Note : this is happening because **'⟬{}⟭'.format(' '.join(correct))** [Join with space character]
   This might be a bug, not sure just wanted to highlight here. 

In addition, 
- Could you please suggest **performance improvements** for **_find_candidates_window_n** method given that all valid words are provided as input for given incorrect token (no need for trie/dictionary based lookup). Due to multiple nested levels of for-loops, its taking huge amount of time when using in practice.

Would like to add improvements as PR 👍

**Comment by yoptar** [comment]
1. I couldn't think of anything better than a simple accuracy when comparing misspellings with the highest error model probability against a correct word. One could also look at top 2 and top 3 corrections.
1. 
1.  Yep, that's a bug, thank you for noticing. I think, I added that to be able to build error models on sentences and not just words, but it did not work then and I just left it there... It also looks like it's not the only problem, because I still get scores worse than on a pretrained error model for Russian language that one can download form our servers,  even after the spaces are removed. I'll try to figure it out, thanks again.
1. I couldn't find other ways to improve the algorithm itself when I wrote it. One can try and rewrite it using Cython or even raw C, it should add some boost in performance, but I'm not sure if it's worth it.

**Comment by yoptar** [comment]
I'll assume that the issue is resolved for now. Feel free to reopen it or create a new one if you have any further questions.

**Comment by smilenrhyme** [comment]
@yoptar Thanks for overwhelming response :) 

Few more points, after using in practice

- Should we do inference only on max-edit=2 in error model as we are training with same max distance ?
- What is intuition behind having this threshold : **prop_threshold=1e-6** and how its obtained.
- The output score which comes out at spell correction level is log_prob. If we convert into usual probability, the output_prob is very very small (of order of 10 ^ -28). Given this range, how do we compute **model operating point** (with some precision & recall). Any thoughts.
- Could you suggest some hints w.r.t  **conditioning on the position** of the
edit relative to the source word (which is mentioned in noisy channel error model paper) w.r.t current code e.g. {start_of_word, middle_of_word, end_of_word}
- Is this useful https://nlp.stanford.edu/pubs/toutanova2002spelling.pdf (modelling phonetic errors) ?

Thanks a lot once again 👍

**Comment by yoptar** [comment]
- if I remember correctly, we can use any max-edit smaller or equal then the one we used in training, as "training" itself is just a statistics calculation.
- `1e-6` was set empirically to speed up the trie search a little without drastically affecting scores on our data. but it should have been an initialization parameter.
- I don't know what **model operating point** is, sorry
- For this I've added special characters to mark word's beginning and end (`⟬` and `⟭`) so that they would be taken into account in the model.
- sure, I think that phonetics should be used as a feature in a perfect spelling correction algorithm, but it can also slow everything down even more :slightly_smiling_face:

**Comment by smilenrhyme** [comment]
@yoptar Thanks a lot :)

**Comment by smilenrhyme** [comment]
@yoptar Found some more points for discussion, 

1. Here candidate is getting split into words. Shouldn't we query LM directly without split ? (pros/cons)
https://github.com/deepmipt/DeepPavlov/blob/c10b079b972493220c82a643d47d718d5358c7f4/deeppavlov/models/spelling_correction/electors/kenlm_elector.py#L67

2. Here we are overriding error probability to its max value, that means only LM score will going to play role into final ranking. Is there any better way of handling this.
https://github.com/deepmipt/DeepPavlov/blob/c10b079b972493220c82a643d47d718d5358c7f4/deeppavlov/models/spelling_correction/brillmoore/error_model.py#L138

Thanks a lot for the throughout support 👍

**Comment by smilenrhyme** [comment]
@yoptar please share your thoughts for above points. Thanks a lot !!

**Comment by yoptar** [comment]
Hi @smilenrhyme,

1. the LM wont know how to score two or more tokens disguised as one, it wasn't trained on such data. This is done for cases where we assume spaces were skipped in a correct sentence.
2. We do not change this word so it doesn't really matter what probability to it, only LM score will fluctuate depending on a beam in the beam search.

**Comment by smilenrhyme** [comment]
Thanks @yoptar :) 

Just to clarify point number 1, is this true even when LM(KenLM) was trained with order 4 and each training instance was "a sentence including spaces" per line.

**Comment by smilenrhyme** [comment]
@yoptar please share your views on above point. Thanks a lot :)

**Comment by yoptar** [comment]
Hi @smilenrhyme,

I think the order was 3. And yes, there was a sentence per line with spaces included

**Comment by smilenrhyme** [comment]
@yoptar if it is trained on such data then should not we query LM directly without split into words ?

Thanks !!

P.S Sorry for asking too many questions.

**Comment by yoptar** [comment]
Tokenization is included in the training process. We're using a low level method for calculating current score, and it requires a single token as its input.
You can try it and prove me wrong, of course. :slightly_smiling_face:

**Comment by smilenrhyme** [comment]
@yoptar Yes its at word level, just went through the code documentation. Below is the example but not sure what is the difference though. (It does not throw any error)

```
In [60]: s1 = kenlm.State()
In [61]: s2 = kenlm.State()
In [62]: model.BeginSentenceWrite(s1)
In [63]: accum = 0.0
In [64]: accum += model.BaseScore(s1, "a", s2)
In [65]: accum += model.BaseScore(s2, "sentence", s1)
In [66]: accum += model.BaseScore(s1, "</s>", s2)
In [67]: accum
Out[67]: -9.504997849464417
```

```
In [68]: accum = 0.0
In [69]: s1 = kenlm.State()
In [70]: s2 = kenlm.State()
In [71]: model.BeginSentenceWrite(s1)
In [72]: accum += model.BaseScore(s1, "a sentence </s>", s2)
In [73]: accum
Out[73]: -8.628549575805664
```

**Comment by smilenrhyme** [comment]
@yoptar Got the answer, In the second case its "unknown" probability. It does not change even we add tokens to that. 

Thanks for your time and support :) 👍

**Comment by smilenrhyme** [comment]
@yoptar 

One suggestion to improve on current strategy, in below code snippet we are iterating over all prefixes. Can we use **_LevenshteinSearcher_** style to lookup candidates up to certain edit distance with combination of error model probabilities.

https://github.com/deepmipt/DeepPavlov/blob/c10b079b972493220c82a643d47d718d5358c7f4/deeppavlov/models/spelling_correction/brillmoore/error_model.py#L110

Please share your thoughts. Thanks 👍

**Comment by smilenrhyme** [comment]
@yoptar In addition to above, 

Could you please suggest some pointers/code changes required to support "Phrase based" error probabilities. I have come across [this](https://pdfs.semanticscholar.org/7bd7/65846bbab248e45d74891f42e6ae8f308874.pdf) but not sure how this can be fit in current settings.

Thanks a lot :)

**Comment by smilenrhyme** [comment]
@yoptar Please provide your useful comments on above questions, Thanks a lot !!

**Comment by smilenrhyme** [comment]
@yoptar please provide your views. It would be a lot of help for us. Thanks.

**Comment by yoptar** [comment]
Hi @smilenrhyme,
I'm sorry, but I didn't read that paper. I haven't done anything in spelling correction for some time so I'm not really up to date in that field. We do have a new spelling correction model in development, but I don't know a lot about it yet. I suggest you ask this question in our [forum](http://forum.deeppavlov.ai/).

**Comment by smilenrhyme** [comment]
Thanks a lot 👍
